stix2/base.py

```diff
@@ -347,6 +360,61 @@ def _check_property(self, prop_name, prop, kwargs):
             for ref in kwargs[prop_name]:
                 self._check_ref(ref, prop, prop_name)
 
+    def _generate_id(self, kwargs):
+        required_prefix = self._type + "--"
+        namespace = uuid.UUID("00abedb4-aa42-466c-9c01-fed23315a9b7")
+
+        try:
+            properties_to_use = self._id_contributing_properties
+            if properties_to_use:
+                streamlined_object = {}
+                if "hashes" in kwargs and "hashes" in properties_to_use:
+                    possible_hash = self._choose_one_hash(kwargs["hashes"])
+                    if possible_hash:
+                        streamlined_object["hashes"] = possible_hash
+                for key in kwargs.keys():
+                    if key in properties_to_use and key != "hashes":
+                        if type(kwargs[key]) is dict:
+                            for otherKey in kwargs[key]:
+                                if isinstance(kwargs[key][otherKey], _STIXBase):
+                                    streamlined_object[key] = self._embed_obj_to_json(kwargs[key][otherKey])
+                                else:
+                                    streamlined_object[key] = kwargs[key]
+                        else:
+                            if isinstance(kwargs[key], _STIXBase):
+                                streamlined_object[key] = self._embed_obj_to_json(kwargs[key])
+                            else:
+                                streamlined_object[key] = kwargs[key]
+
+                if streamlined_object:
+                    data = canonicalize(str(streamlined_object), utf8=False)
+                    return required_prefix + str(uuid.uuid5(namespace, str(data)))
```

Comment:
JSON is text.  So `str(data)`, under python3 is redundant.  `data` should already be a string.  Under python2, this will try to encode data using some default encoding, which is likely to fail if there are high-codepoint characters in it.  Looks like the uuid.uuid5() implementation already encodes the given name to UTF-8, so you don't have to do it.  Just pass the name string directly.

Reply:
I think when I originally prototyped this - data was a bytearray not a string...

Reply:
I hadn't realized that this code had been mostly lifted right out of your prototype.  I went back and found it again, and you just had `canonicalize(streamlined_object)`, without the utf8=False part.  So yeah, it would have returned bytes.